include/rocksdb/comparator.h

```diff
@@ -37,7 +38,17 @@ class Comparator {
 
   virtual ~Comparator() {}
 
+  static Status CreateFromString(const ConfigOptions& opts,
+                                 const std::string& id,
+                                 const Comparator** comp);
   static const char* Type() { return "Comparator"; }
+  static const char* kBytewiseClassName() {
+    return "leveldb.BytewiseComparator";
+  }
+  static const char* kReverseBytewiseClassName() {
+    return "rocksdb.ReverseBytewiseComparator";
+  }
```

Comment:
Should not these be defined in derived class?

Reply:
They are defined here so they can be used publicly.  If they are in the derived class, then these constants cannot be used as strings/inputs to CreateFromString.  Generally speaking, I have thought the names of the builtin Customizable objects should be public, but I can go either way.

Reply:
It looks odd to me that the derived class name is defined in base class. How about other compactors?

Reply:
So, the sampling is small (so far) as there are not that many Customizable classes.  

For TableFactory, the derived names (Block, Plain, Cuckoo) are in the base class.  This is necessary because there is often casting that goes on based on the type and the derived classes are not in the public API.

For Env, I could see the potential of needing the "default" or "posix" names in the base class.  I am not certain that other names will have to be there.  Default might be necessary so that one could cast an Env to the default or base implementation.

If the name is not in the base class, then constructing a derived class can be more problematic.  The options are:
- Create a unique constructor for every type (e.g ByteWiseComparator(), ReverseByteWiseComparator(), etc).  This is obviously not sustainable for all permutations
- Document what the names are to create each type (e.g. "Comparator has two built-in types.  a "leveldb.BytewiseComparator" and a "rocksdb.ReverseBytewiseComparator".  These names can be passed to CreateFromString to create ...").  This is changing the problem from having a constant in the public API to having the behavior in documentation.
- Make the implementation classes public.  Yuck.

I am not opposed to making it a "documentation" problem.  I have also thought about ways of creating a tool that would allow a user to see what "instances" are available (via the ObjectRegistry) to discover what these names are and possibly make the features self-documenting.

Since the names (for Comparators) are not really necessary in the public API at the moment, I can move them to the implementation classes if that is better.